apps/sms/views/shared/js/utils.js

```diff
@@ -711,6 +711,57 @@
     },
 
     /**
+     * Returns a rate limited function, useful to reduce the number of function
+     * calls when listening to user input (keypress) or browser events.
+     * @params {function} func - The function that should be rate-limited
+     * @params {number=300} delay - Min amount of time (ms) between two calls
+     * @param {Boolean=false} preventfirstCall - Prevent first synchronous call 
+     * when true.
+     * @param {Boolean=false} preventLastCall - Prevent last trailing call
+     * when true.
+     * @returns {function} - The rate limited function that will call the
+     *  original function
+     */
+    throttle: function(func,
+                       delay = 300, 
+                       preventfirstCall = false, 
+                       preventLastCall = false) {
+      var timeout = null;
+      var previous = 0;
+
+      if(typeof func !== 'function'){
+        throw new Error('func must be a Function');
+      }
+
+      if(typeof delay !== 'number' || delay < 0){
+        throw new Error('delay must be a positive number');
+      }
+
+      return function(...args) {
+        var now = Date.now();
+        if (!previous && preventfirstCall){
+          previous = now;
+        }
+        var remaining = delay - (now - previous);
+        if (remaining <= 0) {
+          if (timeout) {
+            clearTimeout(timeout);
+            timeout = null;
+          }
+          previous = now;
+          func.apply(this, args);
+        } else if (!timeout && !preventLastCall) {
+          timeout = setTimeout(() => {
+            previous = preventfirstCall ? 0 : Date.now();
```

Comment:
Sorry but by "reserved", do you mean that the line 742 condition is not reached ?